REMOVED_SETTINGS

juj opened the pull request:
Make the compiler silently allow removed -s settings, if their value aligns with behavior that is now the only supported option.

Without this, whenever an option is removed, it destroys backwards compatibility in build systems.  See https://github.com/emscripten-core/emscripten/pull/7422#issuecomment-461417994

sbc100 commented on src/settings.js:
```diff
@@ -1419,3 +1419,17 @@ var TARGET_BASENAME = '';
+// [OPTION_NAME, POSSIBLE_VALUES, ERROR_EXPLANATION], where POSSIBLE_VALUES is an array of values that will
+// still be silently accepted by the compiler. First element in the list is the canonical/fixed value going forward.
+// This allows existing build systems to keep specifying one of the supported settings, for backwards compatibility.
+var REMOVED_SETTINGS = [
```
Can we move this directly into the shared.py python code?  There is no need to for REMOVED_SETTINGS to be part of settings.js is there?

juj replied:
Having these in src/settings.js is intentional for two reasons. First one is that people currently know to look in src/settings.js for options, and they rely on the documentation in there. Having these in src/settings.js helps people discover them to figure out which settings have gone away. Also, having the options in src/settings.js helps the process of removing options (just move them in the array at the end of the same file).

If the info related to deprecated options was in a separate file than the options themselves, users will not discover them easy, and it can be easier to forget to update the list when removing options.

sbc100 replied:
But the comment right above already says "Legacy settings that have been removed...".

I would argue that this CL effectively means these settings are still supported (albeit with a fixed value).  IIUC, they will also still be supported in user code too.   User JS code can still check `SPLIT_MEMORY` right?   They are settings that we are carrying over in some form from an earlier time, to me this what legacy means.

juj replied:
This is getting to odd name bikeshedding. The options were first legacy, and then the options have now been removed. I don't think it makes sense to argue over "is a setting a setting if its value is now fixed". I'll just change the name.

sbc100 commented on tools/shared.py:
```diff
@@ -1224,6 +1227,11 @@ def __getattr__(self, attr):
         raise AttributeError
 
     def __setattr__(self, attr, value):
+      for legacy_attr, fixed_values, error_message in self.attrs['REMOVED_SETTINGS']:
```
is we move REMOVED_SETTINGS into this python file we can make a map an you can do:

```
if attr in REMOVED_SETTINGS:
  fixed_values, error_message = REMOVED_SETTINGS[attr]
  if value not in fixed_values:
    logger.error('Invalid command line option -s ' + attr + '=' + str(value) + ': ' + error_message)
    sys.exit(1)
```

Also, why not issue a warning an return here.  There no point in actually applying the setting is there?

Or are you trying to handle the case where user JS code is reading these attributes?

juj replied:
Applying the settings is needed, since these become options that user code can `#if` against. User code in their own JS library could have had e.g. `#if MEMFS_APPEND_TO_TYPED_ARRAYS` to have a custom block of code that depends on a setting. If we no longer apply these, such user code can silently break without an error being issued.

trzecieu commented:
Hi @juj,
Thank you for taking a look at this. There is no rush at all for handling this, my comment was rather a feedback from one of use cases. 
This implementation although supper explicit, it will add extra maintenance overhead to remember to populate REMOVED_FLAGS array. I share also @sbc100's opinion that settings,js might not be the best place for such array, as elements of this object can be overridden with `-s` flag.

juj commented:
> There is no rush at all for handling this, my comment was rather a feedback from one of use cases.

No worries. I do this since our build system also regressed with the recent removal of `MEMFS_APPEND_TO_TYPED_ARRAYS` option. We have been using `-s `MEMFS_APPEND_TO_TYPED_ARRAYS=1` option, and now that has become the only possible option, but our build system can no longer pass that flag, so we must remove it to build with new Emscripten; but we can then no longer build against older Emscripten, because we need the `-s `MEMFS_APPEND_TO_TYPED_ARRAYS=1` flag. We have two teams that update Emscripten on a different cadence, hence, problems.

So treating missing `-s` options like an error is definitely a bidirectional breakage barrier, and we need to fix it.

kripken commented on tools/shared.py:
```diff
@@ -1224,6 +1227,11 @@ def __getattr__(self, attr):
     def __setattr__(self, attr, value):
+      for legacy_attr, fixed_values, error_message in self.attrs['REMOVED_SETTINGS']:
+        if attr == legacy_attr and value not in fixed_values:
+          logger.error('Invalid command line option -s ' + attr + '=' + str(value) + ': ' + error_message)
```
this error is reporting that a legacy key is used with the wrong value, I think? if so, the error message could be more specific on that, and report the only currently supported value?

juj replied:
I wanted to keep this side of the print very unstructured, so that the actual "meat" of the print will then be possible to freely formatted in src/settings.js side, depending on what is desired to be said about each item. If you combine this string with those in src/settings.js, I hope the prints should look readable?

kripken commented on tools/shared.py:
```diff
@@ -1224,6 +1227,11 @@ def __getattr__(self, attr):
+        if attr == legacy_attr and value not in fixed_values:
+          logger.error('Invalid command line option -s ' + attr + '=' + str(value) + ': ' + error_message)
+          sys.exit(1)
+
```
if the value *is* the proper value, so we do not error here, it might be nice to issue a warning, which would suggest to people to remove it. otherwise this PR would make us no longer give any indication about options we remove (unless they have a different value)?

sbc100 replied:
Yes, I think its important that developers get warning around use of legacy options.

In addition, if we do go with such warnings then there is less argument for keeping REMOVED_SETTINGS in the js file, since all users will be made aware of such legacy usage without needing to see the contents of the js file.

juj replied:
No, we do not want to issue a warning in this case, because it will hint developers towards a direction that breaks backwards compatibility. There is no benefit in breaking backwards compatibility for the joy of breaking backwards compatibility. We want to guide developers to break backwards compatibility when there is a win to be had from doing so.

Developers who deliberately want to keep supporting multiple Emscripten versions would have the annoyance to have to live with spurious warnings being emitted by the compiler when they are doing the right thing, or build that clumsy network of `emcc -v` version sniffing, to avoid the "to get X you cannot say X" issue.

sbc100 replied:
Personally, as a developer, when I update to a new compiler version, I do like to know when the compiler options I'm using are legacy/removed, so that I can cleanup and modernize my build.  Especially, as is the case with `-s` settings, this could mean I can cleanup `ifdef` regions of my code.   But you are right, that only really works if I don't plan to support older versions.

sbc100 commented:
I'm not a big fan having the legacy settings continue to work.  I think I would rather error out with a meaningful message.  

That way the developer can take action and remove the flag from their build system.

Developers already have to deal with different flags from different compiler versions.  Its part of the compiler upgrade process.   Am I missing something?  Is this not a normal and expected part of switching compiler versions?

Pretending to support flags that we no longer actually do seems like it could also be doing the developers a disservice.  

Perhaps a compromise would be to keep this solution around until we do a major version bump?    In which case I propose we do one soon :)

juj commented:
> Yes, I think its important that developers get warning around use of legacy options.

That is not what is happening here. This is not about suppressing the use of removed -s options, but this is about suppressing the use of removed -s options with wrong values that are no longer supported.

For example, we used to have an option `-s MEMFS_APPEND_TO_TYPED_ARRAYS=0/1`, where 0 meant that filesystem files would use regular JS arrays to append content to them, and 1 would mean that typed arrays (with geometric grow semantic) would be used to append bytes to files.

People then go and add linker settings -s MEMFS_APPEND_TO_TYPED_ARRAYS=0 or -s MEMFS_APPEND_TO_TYPED_ARRAYS=1 to their projects.

Then we dropped support for option -s MEMFS_APPEND_TO_TYPED_ARRAYS=0, keeping the typed array scenario as the only supported case.

But because we deleted the -s MEMFS_APPEND_TO_TYPED_ARRAYS= option, any developers who were passing the linker setting -s MEMFS_APPEND_TO_TYPED_ARRAYS=1, would now have their build broken; even though they have opted in to the exact behavior that we are enforcing that everyone should use. There is absolutely no reason to break users when they pass settings that are exactly like we want them to be.

Anyone setting -s MEMFS_APPEND_TO_TYPED_ARRAYS=0 should still keep getting an error. Anyone setting -s MEMFS_APPEND_TO_TYPED_ARRAYS=1 should be silently accepted, since they are explicitly saying they are doing what compiler wants them to.

There is even no reason to issue a build warning about it, except perhaps in verbose/debug builds. That is just needless clutter for a scenario where developer is opting in to an explicit contract that compiler also requires. It is like party A saying "I want X", and party B replying "We only support X, but you cannot ask for X, you'd get X only if you do not ask for it, but you did, so get lost."

The problem with this is that your build system immediately breaks when the setting is removed, even if your build system had been using the desired `-s MEMFS_APPEND_TO_TYPED_ARRAYS=1` behavior. So you cannot upgrade the compiler unless you also upgrade the build system to drop `-s MEMFS_APPEND_TO_TYPED_ARRAYS=1` option. And after you upgrade the build system and the compiler, you have just lost compatibility with older compilers and cannot use that build system anymore with any compiler version where `-s MEMFS_APPEND_TO_TYPED_ARRAYS=1` did not happen to be the default. Which for this option means that all compilers older than January 23rd 2019 are now broken for you. So all projects that are using that build system are forced to upgrade at the same time.

Or you spend time to do magic behavior guessing against the compiler, doing test builds with `emcc tests/hello_world.c -s MEMFS_APPEND_TO_TYPED_ARRAYS=1` to detect if compiler wants to hear about `-s MEMFS_APPEND_TO_TYPED_ARRAYS=1` to enable `-s MEMFS_APPEND_TO_TYPED_ARRAYS=1` or not. The complexity is ridiculous.

And this all happens to you even if none of your projects were never even remotely interested with using `-s MEMFS_APPEND_TO_TYPED_ARRAYS=0` option for more than 3 years now that the setting has existed. You just get free churn to do pointless build system plumbing that wastes developer time and makes build systems more complex.

These kind of bidirectional breaking barriers create immediate chicken and egg problems in multi-team setups, where everyone needs to update in lockstep, and the result is that the update will be deferred until much later when everyone is able to work in sync.

Just because you could not ask for X to get X.

If we remove an option that has a clear specific value that is now the only accepted value, then that one value needs to be silently accepted, even if the setting itself was dropped. Attempting to change the setting to any other value is fine to cause compilation to abort.

sbc100 commented:
OK, you have made some good arguments.  Enough that I think we can move forward with this change in some form.

I can see that continuing to support `-s MEMFS_APPEND_TO_TYPED_ARRAYS=1` in the short term is useful, and you've made good arguments for it.   I hadn't realized this was only for certain options (where we remove the choice but kept the thing itself) and not all removed options.

Can we agree to eventually remove these completely down the line some time?  Perhaps on a major version upgrade?  Otherwise it leave cruft in the developers build system (they will never remove the redundant option) and in emscripten itself.

juj commented:
> Can we agree to eventually remove these completely down the line some time?

The maintenance cost for leaving an old entry around is exactly one line in src/settings.js, given that the machinery is generic.

> Otherwise it leave cruft in the developers build system (they will never remove the redundant option)

There is no problem with a build system having redundant options, and one can argue that the option is not redundant, if the build system wants to keep the support for targeting an older Emscripten version as well. If we remove the option even after time passed, it is still an explicit breakage that we cause; and the cost of avoiding that breakage would have been as simple as keeping one line around in src/settings.js. That registry also acts as a ChangeLog of sorts as well, documenting different historical options and their current values, which is nice.

>  this was only for certain options (where we remove the choice but kept the thing itself)

This is actually the most common way that these options go away - one of the choices becomes the only supported thing. So this kind of "deleted values" registry is expected to be the default way of handling removed options.

sbc100 commented:
It looks like there is another place were we handle such legacy settings:
https://github.com/emscripten-core/emscripten/pull/8092

If we land this change does it makes sense to convert these over to this new system?

sbc100 commented:
Can we rename this from REMOVED_SETTINGS to LEGACY_SETTINGS?  (and give the CL a more meaningful title).

With that change this LGTM.

sbc100 commented on tools/shared.py:
```diff
@@ -1224,6 +1227,11 @@ def __getattr__(self, attr):
+      for legacy_attr, fixed_values, error_message in self.attrs['REMOVED_SETTINGS']:
+        if attr == legacy_attr and value not in fixed_values:
+          logger.error('Invalid command line option -s ' + attr + '=' + str(value) + ': ' + error_message)
+          sys.exit(1)
```
Can you use "exit_with_error()" rather than these two lines?

juj replied:
Sure

juj commented:
>It looks like there is another place were we handle such legacy settings:
> #8092

> If we land this change does it makes sense to convert these over to this new system?

Yes, migrated.

> Can we rename this from REMOVED_SETTINGS to LEGACY_SETTINGS? (and give the CL a more meaningful title).

No, legacy implies something that still exists as opposed to something that has been removed. Saying e.g. that option X is legacy would mean that while not being considered "modern" or "current" anymore, it is still an option that does exist. We want to signal that these options no longer exist and flipping their values will not produce different effects that the option used to do.

sbc100 commented on emcc.py:
```diff
@@ -1025,17 +1025,6 @@ def check(input_file):
     # Use settings
 
-    try:
-      assert shared.Settings.ASM_JS > 0, 'ASM_JS must be enabled in fastcomp'
```
Can we remove the direct entries for these from Settings.js (perhaps also assert that a setting cannot be both at the top level and also in REMOVED_SETTINGS?)

sbc100 replied:
Nice!

sbc100 commented:
I'm sorry for all the back and forth on this PR.  Its has been good to understand you perspective on backwards compat.  Sorry if it seems like I'm nitpicking or disagreeing with deliberately, that has certainly not been my intention.

juj commented:
Thanks, much appreciate the effort to review here, and the goodwill. Happy that we found a good way forward, landing this now.
